Topic/api 3/create appointment post endpoint

Liriaene opened the pull request:
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)



* **What is the current behavior?** (You can also link to an open issue here)



* **What is the new behavior (if this is a feature change)?**



* **Other information**:

Liriaene committed:
API-3: Condensed sql to use more funtional style programming

Liriaene committed:
API-3: Added package and script to debug server with live code changes

Liriaene committed:
API-3: Added appointmentPOST endpoitn, endpoitn now useing db handler

Liriaene committed:
API-3: Updating bcrypt dependencies

gstone271 commented on api/appointmentsPOST.ts:
```diff
@@ -0,0 +1,115 @@
+  ]
+
+  let tfisurvey_sql : string = "INSERT INTO tfisurvey (" + tfi_datapoints.join(", ") + ") VALUES ("+ tfi_datapoints.map((value, index) => "$" + (index + 1)) + ") RETURNING *";  
+  let tfisurvey_values = [patientid, tfi_i, tfi_sc, tfi_c, tfi_si, tfi_a, tfi_r, tfi_q, tfi_e, tfi_overallscore];
```
You can use `map` to transform `tfi_datapoints` into this, by turning `name` into `request.body[name]`, thus eliminating the need for line 10.

gstone271 commented on api/app.ts:
```diff
@@ -25,8 +25,10 @@ app.get('/heartbeat', heartbeatEndpoint);
 app.post('/login', loginEndpoint);
 
-import appointmentsEndpoint from './appointments';
+import appointmentsEndpoint from './appointments'
```
missing ;

gstone271 commented on api/appointmentsPOST.ts:
```diff
@@ -0,0 +1,115 @@
+  let thssurvey_sql : string = "INSERT INTO thssurvey (" + ths_datapoints.join(", ") + ") VALUES (" + ths_datapoints.map((value, index) => "$" + (index + 1)) + ") RETURNING *";
+  let thssurvey_values = [patientid, ths_sectiona, ths_sectionb, ths_sectionc, ths_sectionc_example];
+  
+  let tssurvey_sql : string = "INSERT INTO tssurvey (" + ts_datapoints.join(", ") + ") VALUES (" + ts_datapoints.map((value, index) => "$" + (index + 1)) + ") RETURNING *";
```
You can write a function or loop that does the same thing for all of `tfisurvey_sql`, `thssurvey_sql`, etc. without duplicating code (only possible if you implement the line 85 change).

ElijahRW commented on api/appointmentsPOST.ts:
```diff
@@ -0,0 +1,115 @@
+    rightear_lowf_severity, 
+    rightear_highf_severity, 
+    leftear_lowf_severity, 
+    leftear_highf_severity, rightear_lowf_configuration, rightear_highf_configuration, leftear_lowf_configuration, leftear_highf_configuration, audiogramtype
```
This should be pushed out to `root/api-objects` directory as its own class. (Liam will be referencing this when sending data over the API Service

gstone271 replied:
Per line 85 comment we want to get rid of this class instead.

ElijahRW reviewed (approved):
Looks good. I agree with  Griffin's suggested edits, but there are no pieces that compromise performance or functionality.

Commit:
API-3: Updated sql to nolonger use values object.

Commit:
More changes. They are not tested yet however last time I tried this approch I got a syntax error in my SQL. Hopefully that will not be the case again

Liriaene committed:
API-3: More sql condensing

This is a consolidation of the sql down to one line per call. Some bug fixes I still need to make but this is my current direction

Liriaene committed:
API-3: Removed dead data structure

Liriaene committed:
API-3: Bug fix

Liriaene committed:
API-3: Switch appointment SQL to functional style as well

Liriaene committed:
API-3: Uploading my Postman collection for testing appointment api

Liriaene committed:
API-3: Unused import

gstone271 commented on api/appointmentsPOST.ts:
```diff
@@ -0,0 +1,102 @@
+    'appointmentdatetime'
+  ]
+
+  let tfisurvey_sql : string = "INSERT INTO tfisurvey (" + tfi_datapoints.join() + ") VALUES ("+ tfi_datapoints.map((value, index) => request.body[value]) + ") RETURNING *"; 
```
This is SQL injectable.

You must always use the $1 $2 etc in the SQL string and then have a separate array of values.
Use an additional, separate call to `map` to generate the array of values.
This will also eliminate the need to do stuff differently depending on datatype.
The map you've written (the line 59 version) will work for the sql values array though.

gstone271 commented on api/appointmentsPOST.ts:
```diff
@@ -0,0 +1,102 @@
+  return await db(async (connection) => {
+    // Take data from request and build sql call from i
+
+  let tfi_datapoints : string[]  = [
```
These arrays of constants can be moved out of `handler`

Liriaenes committed:
Updating some of the SQL to use values again.

Liriaenes committed:
Dealing with SQL injectable code, Addeding newline to postman collection.

ElijahRW reviewed (approved):
Could use refactoring in future. (No major security holes)
